File: Problems/Algorithms/376. Wiggle Subsequence/WiggleSubsequence.java
class Solution {
    public int wiggleMaxLength(int[] nums) {
        int n = nums.length;
        if (n < 2) return n;
        
        int diff = nums[1] - nums[0];
        int ans = diff == 0 ? 1 : 2;
        for (int i = 2; i < n; i++) {
            int curr = nums[i] - nums[i-1];
            if ((curr > 0 && diff <= 0) || (curr < 0 && diff >= 0)) {
                ans++;
                diff = curr;
            }
        }
        
        return ans;
    }
}

File: Problems/Algorithms/376. Wiggle Subsequence/wiggle_subsequence.rs
impl Solution {
    pub fn wiggle_max_length(nums: Vec<i32>) -> i32 {
        let (down, up) = nums.windows(2).fold((1, 1), |(mut down, mut up), x| {
            if x[1] > x[0] { up = down + 1 } else if x[1] < x[0] { down = up + 1 };
            (down, up)
        });
        down.max(up)
    }
}

